bindings/matrix-sdk-crypto-ffi/src/verification.rs

```diff
@@ -1,101 +1,365 @@
+    pub fn as_sas(&self) -> Option<Arc<Sas>> {
+        if let InnerVerification::SasV1(sas) = &self.inner {
+            Some(Sas { inner: sas.to_owned(), runtime: self.runtime.to_owned() }.into())
+        } else {
+            None
+        }
+    }
+
+    /// Try to represent the `Verification` as an `QrCode` verification object,
+    /// returns `None` if the verification is not a `QrCode` verification.
+    pub fn as_qr(&self) -> Option<Arc<QrCode>> {
+        if let InnerVerification::QrV1(qr) = &self.inner {
+            Some(QrCode { inner: qr.to_owned() }.into())
+        } else {
+            None
+        }
+    }
+}
+
 /// The `m.sas.v1` verification flow.
 pub struct Sas {
-    /// The other user that is participating in the verification flow
-    pub other_user_id: String,
-    /// The other user's device that is participating in the verification flow
-    pub other_device_id: String,
-    /// The unique ID of this verification flow, will be a random string for
-    /// to-device events or a event ID for in-room events.
-    pub flow_id: String,
-    /// The room ID where this verification is happening, will be `None` if the
-    /// verification is going through to-device messages
-    pub room_id: Option<String>,
-    /// Did we initiate the verification flow
-    pub we_started: bool,
-    /// Has the non-initiating side accepted the verification flow
-    pub has_been_accepted: bool,
-    /// Can the short auth string be presented
-    pub can_be_presented: bool,
-    /// Does the flow support the emoji representation of the short auth string
-    pub supports_emoji: bool,
-    /// Have we confirmed that the short auth strings match
-    pub have_we_confirmed: bool,
-    /// Has the verification completed successfully
-    pub is_done: bool,
-    /// Has the flow been cancelled
-    pub is_cancelled: bool,
-    /// Information about the cancellation of the flow, will be `None` if the
-    /// flow hasn't been cancelled
-    pub cancel_info: Option<CancelInfo>,
+    pub(crate) inner: InnerSas,
+    pub(crate) runtime: Handle,
+}
+
+impl Sas {
+    /// Get the user id of the other side.
+    pub fn other_user_id(&self) -> String {
+        self.inner.other_user_id().to_string()
+    }
+
+    /// Get the device ID of the other side.
+    pub fn other_device_id(&self) -> String {
+        self.inner.other_device_id().to_string()
+    }
+
+    /// Get the unique ID that identifies this SAS verification flow.
+    pub fn flow_id(&self) -> String {
+        self.inner.flow_id().as_str().to_owned()
+    }
+
+    /// Get the room id if the verification is happening inside a room.
+    pub fn room_id(&self) -> Option<String> {
+        self.inner.room_id().map(|r| r.to_string())
+    }
+
+    /// Is the SAS flow done.
+    pub fn is_done(&self) -> bool {
+        self.inner.is_done()
+    }
+
+    /// Did we initiate the verification flow.
+    pub fn we_started(&self) -> bool {
+        self.inner.we_started()
+    }
+
```

Comment:
I think I'll need supports_emoji here (and in the olm.udl)

Reply:
As mentioned in chat, you only get the emoji in the `SasState` if it's supported.